### PR TITLE
Add NNX WeightNorm (update of #4568)

### LINKED ISSUE #4426: NNX WeightNorm
Is there an NNX implementation of the [WeightNorm](https://flax-linen.readthedocs.io/en/latest/api_reference/flax.linen/layers.html#flax.linen.WeightNorm) module as used in Linen? I can't seem to find it in the [normalization layers](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/normalization.html) implemented in NNX.

I'm in the process of writing a new model in NNX which would require them.

### PR DESCRIPTION
**Pull request by samanklesaria**
# What does this PR do?

This PR rebases the changes in #4568 to avoid merge conflicts with the main branch. That PR resolves #4426 by implementing an NNX WeightNorm class. 

The logic implemented here is similar to that of `nnx.SpectralNorm`, but makes a number of difference design choices:

- The code here uses a lazy initialization strategy rather than `SpectralNorm`'s eager one, which makes the code more concise and doesn't require runtime assertions that initialized parameters exist. 
- The code here uses `map_state` rather than doing modifications in place.
- The code here has a `variable_filter` argument (by default `nnx.PathContains('kernel')`) which specifies which part of the wrapped layer to modify. This is present in the linen version as well.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -34,16 +34,6 @@ repos:
         --extra-keys,
         "cell.metadata.executionInfo cell.metadata.id metadata.kernelspec metadata.vscode metadata.colab cell.metadata.executionInfo.user cell.metadata.executionInfo.user_tz cell.metadata.colab",
       ]
-- repo: https://github.com/astral-sh/ruff-pre-commit
```
Comment (vfdev-5) [inline .pre-commit-config.yaml:37]: This change is unrelated to the PR topic, let's revert this change or make it in a separate PR.

File: flax/nnx/__init__.py [hunk 2]
```diff
@@ -31,7 +31,6 @@
 from .graph import GraphDef as GraphDef
 from .graph import GraphState as GraphState
 from .graph import PureState as PureState
-from . import pytreelib as object
```
Comment (vfdev-5) [inline flax/nnx/__init__.py:34]: Are the changes in this file except `from .nn.normalization import WeightNorm as WeightNorm` really needed?

File: flax/nnx/nn/normalization.py [hunk 3]
```diff
@@ -849,6 +849,138 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+      if self.use_scale:
+        scale_shape = tuple(param_val.shape[ax] for ax in feature_axes)
+        scale_path = path + ("scale",)
+        try:
+          scale_state = state[scale_path]
+          scale_value = scale_state.value
+        except KeyError:
+          key = self.rngs.params()
+          scale_value = self.scale_init(key, scale_shape, self.param_dtype)
+          state[scale_path] = nnx.Param(scale_value)
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py]: Can we code this similarly to SpectralNorm?

File: flax/nnx/nn/normalization.py [hunk 4]
```diff
@@ -849,6 +849,138 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+      if self.use_scale:
+        cast_args.append(scale_value)
+
+      final_dtype = dtypes.canonicalize_dtype(*cast_args, dtype=self.dtype)
+      new_val = jnp.asarray(value_bar, final_dtype)
+
+      return nnx.Param(new_val)
+
+    state = nnx.map_state(apply_weightnorm, state)
+    nnx.update(self.layer_instance, state)
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py]: Can we do an inplace changes like in SpectralNorm?
```
    state = nnx.state(self.layer_instance, nnx.Param)
    for path, param in nnx.to_flat_state(state):
      self._spectral_normalize_inplace(path, param, update_stats=update_stats)
```

File: .pre-commit-config.yaml (hunk 1 above)
Comment (samanklesaria) [reply .pre-commit-config.yaml:37]: Whoops- sorry, forgot to remove this.

File: flax/nnx/nn/normalization.py (hunk 3 above)
Comment (samanklesaria) [reply flax/nnx/nn/normalization.py]: It's actually kind of nice to do lazy initialization here instead of the approach in `SpectralNorm`. We don't need to assert the existence of any specific paths, and we only need to modify the `__call__` method rather than both `__call__` and `__init__`. What if I converted SpectralNorm to this lazy approach instead?

Comment (vfdev-5) [reply flax/nnx/nn/normalization.py]: I think try/except in `__call__` can have a cost when called in the loop

File: flax/nnx/__init__.py (hunk 2 above)
Comment (samanklesaria) [reply flax/nnx/__init__.py:34]: Ruff automatically makes these changes using the standard pre-commit hook. The code in main must not have had the pre-commit hook run. Disabling pre-commit hooks runs directly counter to the contribution guide (https://flax.readthedocs.io/en/stable/contributing.html), so I'm disinclined to do it. Perhaps I should start with a PR that just runs pre-commit hooks on main, and then base this PR on it?

Comment (vfdev-5) [reply flax/nnx/__init__.py:34]: There is test job running pre-commit hooks: https://github.com/google/flax/blob/5a2da6d2e35f1aeebfe50c508602c8f936d9e87f/.github/workflows/flax_test.yml#L34

Comment (samanklesaria) [reply flax/nnx/__init__.py:34]: When I try to commit without these changes, I get the following:

Screenshot 2025-10-22 at 10 12 37 AM

Why might this not have occurred in the `flax_test` workflow? Do you see something similar?

File: flax/nnx/nn/normalization.py [hunk 5]
```diff
@@ -842,6 +842,139 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
       self.epsilon,
     )
 
+class WeightNorm(nnx.Module):
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py:845]: We need to expose this class into the docs: https://flax--5043.org.readthedocs.build/en/5043/search.html?q=WeightNorm

File: flax/nnx/nn/normalization.py [hunk 6]
```diff
@@ -842,6 +842,139 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+    self.param_dtype = param_dtype
+    self.variable_filter = nnx.filterlib.to_predicate(variable_filter)
+
+    if use_scale:
+      state = nnx.state(self.layer_instance, nnx.Param)
+      def init_scales(param):
+        feature_axes = _canonicalize_axes(param.ndim, self.feature_axes)
+        scale_shape = tuple(param.shape[ax] for ax in feature_axes)
+        return scale_init(rngs['params'], scale_shape)
+      self.scales = nnx.data({path: init_scales(param) for path, param in nnx.to_flat_state(state)
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py]: Can we define `self.scales` attribute unconditionally, e.g. if use_scale is False as None?

File: tests/nnx/nn/normalization_test.py [hunk 7]
```diff
@@ -323,6 +323,70 @@ def __call__(self, x, *, mask=None):
...
+    variables = linen_model.init(jax.random.key(1), x)
+
+    nnx_model = NNXModel(dtype=dtype, param_dtype=param_dtype, rngs=rngs)
+    nnx_model.dense.kernel.value = variables['params']['dense']['kernel']
+    nnx_model.dense.bias.value = variables['params']['dense']['bias']
+
+    linen_out = linen_model.apply(variables, x)
+
+    nnx_out = nnx_model(x)
+    np.testing.assert_array_equal(linen_out, nnx_out)
```
Comment (vfdev-5) [inline tests/nnx/nn/normalization_test.py:391]: Can we add `self.scales` comparsion when `use_scale=True`?

**Comment by vfdev-5** [comment]
@samanklesaria looks good to me, just few minor things to improve.

**Comment by vfdev-5** [comment]
@samanklesaria mypy is failing now due to the branch: `self.scales = None`.
